lsp-completion.el

```diff
@@ -411,7 +411,7 @@ The MARKERS and PREFIX value will be attached to each candidate."
                      ((or done? result) result)
                      ((and (not lsp-completion-no-cache)
                            same-session?
-                           (listp (cl-second lsp-completion--cache)))
+                           (cl-second lsp-completion--cache))
```

Comment:
I think it's better to change in `same-session?` block before to something like this

``` elisp
(same-session? (and lsp-completion--cache
                    ;; Special case for empty prefix and empty result
                    (or (cl-second lsp-completion--cache)
                        (not (string-empty-p
                              (plist-get (cddr lsp-completion--cache) :prefix))))
                    (equal (cl-first lsp-completion--cache) bounds-start)
                    (s-prefix?
                     (plist-get (cddr lsp-completion--cache) :prefix)
                     (buffer-substring-no-properties bounds-start (point)))))

```

Reply:
I can reproduce it with a non-empty prefix:

```
(483 nil :lsp-items nil :markers
     (483 #<marker
          (moves after insertion)
          at 486 in ownership-table.tsx>)
     :prefix "st")
```

Reply:
Do you know what's the response from the server here? You can check it via lsp trace io.
Should be in format like `[Trace - 10:38:13 AM] Received response 'textDocument/completion - (284)' in 8ms.`

Here is the spec https://microsoft.github.io/language-server-protocol/specifications/specification-current/#textDocument_completion

> Response:
result: `CompletionItem[]` | `CompletionList` | `null`. If a `CompletionItem[]` is provided it is interpreted to be complete. So it is the same as `{ isIncomplete: false, items }`

So if the server returns `[]`, it will be interpreted as `CompletionItem[]` and treat as the server cannot find any more match on this prefix, and the result is complete. Thus even if you type more, there will be no new result, the cache is correct in that case.
if the server returns just `null`, we may have a bug here

Reply:
Log below to generate this cache. It looks like the response came in after the `ta` was typed. The `s` triggered the completion and got cached even though there was pending input.

```
  const ssat = data.sta
```

```
(460 nil :lsp-items nil :markers
     (460 #<marker
          (moves after insertion)
          at 463 in ownership-table.tsx>)
     :prefix "s")
```

```
[Trace - 11:51:45 PM] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
    "uri": "file:///Users/user/Source/myapp/myapp-team-app/src/components/questionnaires/incorporation/form-section/equity/foobar/ownership-table/ownership-table.tsx",
    "version": 435
  },
  "contentChanges": [
    {
      "range": {
        "start": {
          "line": 9,
          "character": 20
        },
        "end": {
          "line": 9,
          "character": 20
        }
      },
      "rangeLength": 0,
      "text": "s"
    }
  ]
}


[Trace - 11:51:46 PM] Sending request 'textDocument/completion - (198)'.
Params: {
  "textDocument": {
    "uri": "file:///Users/user/Source/myapp/myapp-team-app/src/components/questionnaires/incorporation/form-section/equity/foobar/ownership-table/ownership-table.tsx"
  },
  "position": {
    "line": 9,
    "character": 21
  },
  "context": {
    "triggerKind": 1
  }
}


[Trace - 11:51:46 PM] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
    "uri": "file:///Users/user/Source/myapp/myapp-team-app/src/components/questionnaires/incorporation/form-section/equity/foobar/ownership-table/ownership-table.tsx",
    "version": 436
  },
  "contentChanges": [
    {
      "range": {
        "start": {
          "line": 9,
          "character": 21
        },
        "end": {
          "line": 9,
          "character": 21
        }
      },
      "rangeLength": 0,
      "text": "t"
    }
  ]
}


[Trace - 11:51:46 PM] Sending notification 'textDocument/didChange'.
Params: {
  "textDocument": {
    "uri": "file:///Users/user/Source/myapp/myapp-team-app/src/components/questionnaires/incorporation/form-section/equity/foobar/ownership-table/ownership-table.tsx",
    "version": 437
  },
  "contentChanges": [
    {
      "range": {
        "start": {
          "line": 9,
          "character": 22
        },
        "end": {
          "line": 9,
          "character": 22
        }
      },
      "rangeLength": 0,
      "text": "a"
    }
  ]
}


[Trace - 11:51:46 PM] Received response 'textDocument/completion - (198)' in 176ms.
Result: [ ... ]
```

Reply:
I added some logging in and was able to observe that the result had a full list of results, but `nil` still got cached. I've also seen it have input-pending-p be set and no result within `lsp-request-while-no-input` and it not throw when it should have, therefore caching nil.

Reply:
@kiennq based on this description seems like we cache when we were unable to fetch the candidates from the server, is that the case?

Reply:
The cache happens only when we can fetch the candidates from the server. Else it will throw and skip the assignment.
May be we should change the `completed` check to bellow to make sure the `null` response will be treated as incomplete.
``` elisp
(completed (and resp (not (and (lsp-completion-list? resp)
                               (lsp:completion-list-is-incomplete resp)))))
```

I think we've changed this back and forth a few time to work around some language servers problematic response

Reply:
That’s when it *should* happen. But I’m saying it is happening even when canceled. It’s also failing to cache (setting nil) in cases where it did get a response. There is something odd going on.

Reply:
Okay, I believe I see the issues. There are a few:

1. I have `lsp-completion-use-last-result` set to nil. I'm not sure why that impacts `lsp--throw-on-input`, but it does. I would expect that to be orthogonal. So I'm not actually throwing on input, it carries through, returning nil, which ultimately gets cached as nil.
2. The `lsp--while-no-input` used to populate `items` also effectively cares about `lsp-completion-use-last-result` so it will not throw if that's not set to `t`, meaning it will instead return `nil`, which will be cached.

So, can someone explain why `lsp-completion-use-last-result` disables throwing on input (which breaks caching?)?

Reply:
Can you try to change the `(completed ....)` block right bellow the check of `lsp-completion-use-last-result` to 

> ```emacs-lisp
> (completed (and resp (not (and (lsp-completion-list? resp)
>                                (lsp:completion-list-is-incomplete resp)))))
> ```



> So, can someone explain why `lsp-completion-use-last-result` disables throwing on input (which breaks caching?)?

If we don't use the last result, then just let it carry on. The `(complete ...)` block is the one who break caching

Reply:
Okay, got it. Updated, thanks. This seems to fix it for me.

Reply:
Thanks, can you also change the `(same-session? ...)` block to
``` elisp
(same-session? (and lsp-completion--cache
                    ;; Special case for empty prefix and empty result
                    (or (cl-second lsp-completion--cache)
                        (not (string-empty-p
                              (plist-get (cddr lsp-completion--cache) :prefix))))
                    (equal (cl-first lsp-completion--cache) bounds-start)
                    (s-prefix?
                     (plist-get (cddr lsp-completion--cache) :prefix)
                     (buffer-substring-no-properties bounds-start (point)))))
```

Reply:
Done